core/src/main/rust/questdb-jni/src/log.rs

```diff
@@ -0,0 +1,351 @@
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+use core::fmt::Write;
+use std::cell::UnsafeCell;
+use std::fmt;
+
+use dashmap::DashMap;
+use jni::JNIEnv;
+use jni::objects::{GlobalRef, JClass, JMethodID, JStaticMethodID, JString, JValue};
+use jni::signature::{Primitive, ReturnType};
+use jni::sys::{jint, jlong};
+use log::{Level, Log};
+
+use crate::unwrap_or_throw;
+
+#[derive(Clone, Debug)]
+pub struct JavaLog {
+    obj: GlobalRef,
+}
+
+impl JavaLog {
+    fn new(call_state: &CallState, jenv: &JNIEnv, name: &str) -> Self {
+        let j_name = jenv.new_string(name)
+            .expect("could not construct logger name string");
+        let j_name = jenv.auto_local(j_name);
+        let obj = jenv.call_static_method_unchecked(
+            call_state.log_factory_class.as_obj(),
+            call_state.log_factory_get_log_meth,
+            ReturnType::Object,
+            &[JValue::Object(j_name.as_obj()).into()])
+            .expect("io.questdb.log.LogFactory::getLog(String) call failed")
+            .l()
+            .expect("io.questdb.log.LogFactory::getLog(String) didn't return an object");
+        if obj.is_null() {
+            panic!("io.questdb.log.LogFactory::getLog(String) returned null");
+        }
+        let obj = jenv.new_global_ref(obj)
+            .expect("could not create global reference to io.questdb.log.Log object");
+        JavaLog { obj }
+    }
+
+    #[inline]
+    fn call_log_meth(&self, env: &JNIEnv, meth: JMethodID, meth_name: &str, msg: &str) {
+        let lo = msg.as_ptr() as jlong;
+        let hi = lo + msg.len() as jlong;
+        let lo = JValue::Long(lo);
+        let hi = JValue::Long(hi);
+        let res = env.call_method_unchecked(
+            self.obj.as_obj(),
+            meth,
+            ReturnType::Primitive(Primitive::Void),
+            &[lo.into(), hi.into()]);
+        if let Err(e) = res {
+            let throwable = env.exception_occurred().unwrap();
+            if !throwable.is_null() {
+                env.exception_describe().unwrap();
+            }
+            panic!("io.questdb.log.Log::{}(long, long) call failed: {}", meth_name, e);
```

Comment:
It's very strict to panic here - java likes to throw exceptions here and there.

Reply:
I've changed this so that Java exceptions don't panic, but other `jni::Error` instances still panic.